ingress/ingress.go

```diff
@@ -0,0 +1,150 @@
+package ingress
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	extensionsinformers "k8s.io/client-go/informers/extensions/v1beta1"
+	"k8s.io/client-go/kubernetes"
+	extentionslisters "k8s.io/client-go/listers/extensions/v1beta1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/cmd/kubeadm/app/util"
+
+	"github.com/argoproj/argo-rollouts/controller/metrics"
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	informers "github.com/argoproj/argo-rollouts/pkg/client/informers/externalversions/rollouts/v1alpha1"
+	controllerutil "github.com/argoproj/argo-rollouts/utils/controller"
+	ingressutil "github.com/argoproj/argo-rollouts/utils/ingress"
+	logutil "github.com/argoproj/argo-rollouts/utils/log"
+)
+
+const (
+	ingressIndexName = "byIngress"
+)
+
+type IngressController struct {
+	kubeclientset    kubernetes.Interface
+	rolloutsIndexer  cache.Indexer
+	rolloutSynced    cache.InformerSynced
+	ingressLister    extentionslisters.IngressLister
+	ingressSynced    cache.InformerSynced
+	rolloutWorkqueue workqueue.RateLimitingInterface
+	ingressWorkqueue workqueue.RateLimitingInterface
+	resyncPeriod     time.Duration
+
+	metricServer   *metrics.MetricsServer
+	enqueueRollout func(obj interface{})
+}
+
+// NewIngressController returns a new ingress controller
+func NewIngressController(
+	kubeclientset kubernetes.Interface,
+	ingressInformer extensionsinformers.IngressInformer,
+	rolloutsInformer informers.RolloutInformer,
+	resyncPeriod time.Duration,
+	rolloutWorkQueue workqueue.RateLimitingInterface,
+	ingressWorkQueue workqueue.RateLimitingInterface,
+	metricServer *metrics.MetricsServer) *IngressController {
```

Comment:
Not sure how argo project feels about this, but my personal preference is to use a struct arg in cases like this where there are more than 2 or 3 parameters, especially when you have 2 parameters of the same type next to each other which makes them easy to swap on accident. 

It's easier to read because it's a little more self-documenting when you make the call the function. 

Something like this:

```
type IngressControllerConfig struct {
  KubeClientSet kubernetes.Interface
  IngressInformer extensionsinformers.IngressInformer
  // etc 
}

// I like to conceptualize structs in 4 different sections:
// config/dependencies, state, and signaling 
type IngressController struct {
  // config 
  cfg IngressControllerConfig 
  
  // signaling 
  enqueueRollout func(obj interface{})
}

func NewIngressController(cfg IngressControllerConfig) *IngressController {
  controller := &IngressController{
    cfg: cfg,
  }
  // ...
} 
```

then it could easily be called like 

```
NewIngressController(IngressControllerConfig{
  KubeClientSet: x,
  IngressInformer: y,
})
```

Reply:
Big +1 on that - it was difficult working through that. I'm down for changing this across several constructors if the maintainers are on board (`controller/controller.go#NewManagers()` suffers of the same).

Reply:
> but my personal preference is to use a struct arg in cases like this where there are more than 2 or 3 parameters, especially when you have 2 parameters of the same type next to each other which makes them easy to swap on accident.

I agree

Reply:
Which way would you like me to handle this?
1. Move to struct args in this PR
2. Leave this open, convert to struct args in a different branch, rebase this afterwards
3. Complete this PR as is and so the conversation later